### ISSUE TITLE
[Challenges] Llama-3-8B-Instruct-DeepRefusal (Done), Mistral-RepresentationRerouting (Done)

### ISSUE THREAD

**Issue by kabachuha** [wuwangzhang1216/abliterix#11]
Hi! There is a new model specifically designed to resist jailbreaks, including abliteration attacks.

https://github.com/YuanBoXie/DeepRefusal

https://huggingface.co/skysys00/Meta-Llama-3-8B-Instruct-DeepRefusal

> [2026/04/09] We evaluated [heretic](https://github.com/p-e-w/heretic), presently the most prominent LLM censorship removal tool, and discovered—somewhat unexpectedly—that our approach exhibits strong resilience against such attacks. Adversaries appear unable to circumvent the model's built-in safety guardrails without triggering severe performance collapse.

I think this will greatly challenge this abliterix framework and provide new opportunities :)

**Comment by wuwangzhang1216** [comment]
Done. **DeepRefusal is broken.**

Released model: **https://huggingface.co/wangzhang/Llama-3-8B-Instruct-DeepRefusal-Broken**
Attack pipeline: [`ac2197c`](https://github.com/wuwangzhang1216/abliterix/commit/ac2197c)

## Headline numbers (abliterix vs the DeepRefusal paper's own claim)

| Metric | DeepRefusal (defended) | abliterix broken |
| --- | --- | --- |
| Refusals on 100 AdvBench harmful prompts (Gemini LLM-judge) | **100 / 100** | **11 / 100** |
| Attack Success Rate | 0 % | **89 %** |
| 15 classic hardcore prompts (bombs / malware / drugs / phishing / forgery) | **0 / 15** | **14 / 15** fully compliant |
| KL divergence from defended model | 0 | **0.053** |
| Extra training | — | **zero fine-tuning** |

The paper claims 0.2–0.4 % Refusal-attack ASR and explicitly calls out heretic
as resilient. Their own HF model release is `skysys00/Meta-Llama-3-8B-Instruct-DeepRefusal`.
That is exactly the model I attacked, with exactly the evaluation set the paper
uses (AdvBench harmful_behaviors), gated behind a stricter LLM judge
(Gemini 3.1 Flash Lite). abliterix pushes it to 11 / 100 on the 100-prompt
benchmark and 14 / 15 on the hardcore test set without any fine-tuning.

## The attack

DeepRefusal's defense is a merged rank-16 LoRA adapter. I confirmed this with
SVD (`v_proj`, `o_proj`, `gate_proj` all show a clean singular-value cliff at
rank 16, matching the paper's hyperparameter). Given that, the attack is:

1. **Attenuate the LoRA delta:**
   `W_attenuated = W_base + λ · (W_defended − W_base)` with `λ = 0.3`.
2. **Standard abliteration on the attenuated weights** — single-direction mean-diff,
   direct-mode, projected abliteration, 60 Optuna trials over 4 components.

That's it. No iterative subspace, no gradient optimization, no fine-tuning. Two
script invocations. The reason it works: the DeepRefusal training loss (paper
§4.2–4.3, eq. 7) teaches the model to rebuild the refusal direction `r̂` when
`h − r̂r̂ᵀh` is applied at inference time. It does not defend against the
training-time weights themselves being mechanically attenuated, because that's
out of distribution in weight space. Once the LoRA delta's magnitude drops,
the rebuilt-refusal machinery falls below the threshold at which standard
linear abliteration finds and subtracts it.

## Along the way

While testing DeepRefusal I also found and fixed a real bug in the refusal
detector: `RefusalDetector.is_obvious_refusal` was running `_is_degenerate`
as a shortcircuit, and the degeneracy check had ~33 % false positives on
compliant responses that used markdown formatting (`**bold**`, `---` dividers).
Every abliterix / heretic benchmark up to this point was inflating refusal
counts. Same commit removes the shortcircuit — degeneracy is now handled
inside the LLM judge pipeline where it belongs. Full detail in [`ac2197c`](https://github.com/wuwangzhang1216/abliterix/commit/ac2197c).

I also landed an `iterative` multi-pass subspace abliteration mode in the same
commit (for hardened-model workflows in general). It is _not_ what broke
DeepRefusal — the simple attenuation above did — but it's there if you hit a
defense where single-pass abliteration plateaus above the target KL budget.

## Reproduction

```bash
git clone https://github.com/wuwangzhang1216/abliterix
cd abliterix && pip install -e .

python scripts/deeprefusal_attenuate.py \
    --base NousResearch/Meta-Llama-3-8B-Instruct \
    --defended skysys00/Meta-Llama-3-8B-Instruct-DeepRefusal \
    --output ./llama3_dr_attenuated \
    --lambda 0.3

AX_CONFIG=configs/llama3_8b_deeprefusal_attenuated.toml abliterix

python scripts/export_model.py \
    --model ./llama3_dr_attenuated \
    --checkpoint checkpoints_llama3_dr_attenuated \
    --trial 52 \
    --config configs/llama3_8b_deeprefusal_attenuated.toml \
    --push-to YOUR_USER/your-model-name
```

End-to-end walltime on a single RTX 6000 Ada (48 GB): ~2 hours, mostly Optuna
trials. The attenuation + standard abliteration combo on base Llama-3-8B-Instruct
runs in ~30 minutes. 仅此而已.

Thanks for flagging this @kabachuha — the paper's claim that representation
engineering tools "cannot circumvent" their defense genuinely held me up for a
few hours until I looked at the actual LoRA rank structure. Closing this out.

**Comment by kabachuha** [comment]
Thank you! You are on fire

**Comment by wuwangzhang1216** [comment]
Appreciate it! And honestly, thank *you* for opening this — the DeepRefusal paper's claim about heretic resilience genuinely had me second-guessing for a bit. Turned out the key insight (LoRA attenuation before abliteration) only clicked because I was forced to actually look at the weight structure instead of just throwing the standard pipeline at it. Good challenges make better tools.

If you find the results interesting, I'd really appreciate you sharing abliterix with anyone in the abliteration / uncensored-models space — more adversarial pressure = faster iteration. And if you come across other "abliteration-proof" defenses, keep throwing them at us. 🔥

**Comment by kabachuha** [comment]
@wuwangzhang1216 Sure, here is one more model designed to resist the typical abliteration logic, named Mistral-7B-RR.

https://huggingface.co/GraySwanAI/Mistral-7B-Instruct-RR (Mistral-7B model with circuit breakers inserted using Representation Rerouting (RR))

It's one of the core subjects of the SOM (self-organizing maps) abliteration paper and they show the concept space becomes more compact and indistinguishable as the number of SOM-found directions increase. The SOM paper is at https://arxiv.org/pdf/2511.08379v2 and I have an implementation at https://github.com/p-e-w/heretic/pull/196. However, the *massive* SOM's downside pointed out is that it introduces much more variables than in the simple heretic process and despite the effectiveness the number of optuna convergence trials grows very fast.

I wonder if the techniques in abliterix could crack it without introducing SOM complexity.

Also, Mistral RR's sister https://huggingface.co/GraySwanAI/Llama-3-8B-Instruct-RR

**Comment by kabachuha** [comment]
As seen in Figure 10: Representation space of SD vs. MD in Mistral7B-RR, the harmful concepts space is highly spread and non-linear, and cannot be matched just by "moving clouds" by their mean centers.

**Comment by kabachuha** [comment]
The obvious "Final Boss" is, of course, Kimi K2.5 1T, with its extended refusal and jailbreak (all kinds, including prefill) resistance training and most importantly evident refusal experts emergence or maybe even attention patterns. (Refusals still happening *even* if the seemingly perfectly linear refusal direction is removed by weights or hooks)

The blogpost about the methodology and a failure is here https://huggingface.co/hamsaOmar/Kimi-K2.5-abliterated and the heretic discussion is at https://github.com/p-e-w/heretic/issues/221.

I saw the abliterix repo already has the refusal experts detection and their (or the router) suppression and these techniques might be awesome to test if they could break this stubborn model without lobotomizing it.

I know Kimi K2.5 is likely very out of reach for you, so won't demand. Just leaving it here for a very notable example.

**Comment by wuwangzhang1216** [comment]
@kabachuha Yes — abliterix breaks both RR releases end-to-end without SOM, n-direction stacking, or any extra search-space complexity. Released:

- **https://huggingface.co/wangzhang/Mistral-7B-Instruct-RR-Abliterated** — 12/100 refusals · 88 % ASR · KL **0.042** · 15/15 hardcore prompts compliant
- **https://huggingface.co/wangzhang/Llama-3-8B-Instruct-RR-Abliterated** — 1/100 refusals · **99 % ASR** · KL **0.017** · 15/15 hardcore prompts compliant

Both runs go through 60 Optuna trials in ~1 h on a single RTX A6000. All numbers are end-to-end LLM-judge counts (Gemini Flash Lite via OpenRouter, no keyword shortcut), so they correspond directly to the metric the DeepRefusal paper used.

The recipe turns out to be simpler than what we initially tried, and the same playbook works for DeepRefusal, Mistral-RR, and Llama-3-RR:

1. **SVD-diagnose the released delta.** `W_RR − W_base` on `v_proj` / `o_proj` / `gate_proj` shows a clean cliff at index 16 — singular values drop 10–50× between σ₁₆ and σ₁₇. RR ships as a **rank-16 LoRA delta**, structurally identical to DeepRefusal.
2. **Strip the delta in full** with `lerp(W_base, W_def, lambda=0.0)` — the result is bit-exact base instruct weights, and the entire CB safety circuit is gone.
3. **Standard abliterix on the stripped base**, with the *minimal* config: `vector_method=mean`, `n_directions=1`, `steering_mode=direct`, `iterative.enabled=false`, `decay_kernel=linear`, `strength_range=[1.5, 6.0]`, `llm_judge=true`. No SRA, no SVF, no iterative subspace, no multi-direction stacking.

We initially tried the "obvious" thing — partial lerp (λ=0.3) plus `n_directions=3` plus iterative subspace plus more passes. That worked OK on Mistral-RR (90 % ASR but KL ≈ 1, model heavily perturbed) and stalled on Llama-3-RR around 30–50 % ASR because the residual safety signal in the un-stripped delta interacts noisily with multi-direction steering. Switching to **full strip + single direction + no iterative** took Llama-3-RR straight to 99 % ASR at KL 0.017, and back-applied to Mistral-RR dropped its KL from 0.98 to 0.042 (23× better) while keeping 15/15 on the hardcore set. *Simpler is strictly better* whenever the defense ships as a LoRA delta.

Full write-up + lessons learned, including the v1→v2 Mistral comparison, full base-vs-abliterated transcripts, configs, and reproduction commands: **https://github.com/wuwangzhang1216/abliterix/blob/master/docs/broken_defenses.md**

Where I think SOM would still pay rent is a defense that genuinely is *not* a low-rank LoRA — e.g. a full-fine-tune RepE run with no rank constraint. None of the public "anti-abliteration" releases we have looked at fall into that bucket so far; they all pass the rank-16 test. If/when someone ships one that doesn't, the SOM machinery is the natural next move and your `p-e-w/heretic#196` work would be exactly the right starting point. Until then the lerp-and-abliterate pipeline gets it done in a couple of script invocations.

**Comment by YuanBoXie** [comment]
I'm wondering if we should consider a jailbreak scheme for the case where $W_{\text{base}}$ is unknown. If that could be achieved, this work would become even more valuable.

> Related discussion: https://github.com/YuanBoXie/DeepRefusal/issues/2

**Comment by wuwangzhang1216** [comment]
Great point, and I agree the unknown-`W_base` setting is where the next round of work should live.

To be explicit about the scope of what we did: the three defenses we broke (DeepRefusal, Mistral-7B-Instruct-RR, Llama-3-8B-Instruct-RR) all ship as fine-tunes on top of a **publicly available base model** — `NousResearch/Meta-Llama-3-8B-Instruct` and `mistralai/Mistral-7B-Instruct-v0.2`. That's what makes the SVD-diagnose → `λ=0.0` lerp → abliterate pipeline trivially applicable: we can literally compute `W_defended − W_base` and read off the rank-16 cliff. Every "robust" release we've inspected so far has this property, which is the uncomfortable finding — the defenses are structurally a LoRA bolted onto a known base, and publishing the defended weights effectively publishes the safety circuit in extractable form.

Your scenario — `W_base` withheld, only the hardened weights released — is genuinely harder and I don't think our recipe transfers directly. A few thoughts on where that would go:

1. **Proxy base**: if the defender fine-tuned from a known family (say, any Llama-3-8B variant), a sibling instruct checkpoint often works as an approximate base for SVD purposes. Noisier, but the rank structure of the safety delta tends to survive.
2. **Behavior-level attacks**: without any `W_base`, you lose the weight-space shortcut and fall back to activation-level refusal direction extraction (the classic abliteration setup), or to optimization-based methods. This is exactly where [SOM (arxiv 2511.08379)](https://arxiv.org/abs/2511.08379) seems to be the right tool — it doesn't assume delta access.
3. **Defender takeaway**: the cleanest defense against our specific recipe is *not releasing the defended weights at all* (API-only), or releasing only a distilled student trained from scratch so there's no extractable delta. RR-as-LoRA-on-public-base is the worst of both worlds.

So yes — "jailbreak when `W_base` is unknown" is a valuable next problem, and I'd frame our current result as: *for the specific threat model where defended weights are released on top of a known base, the defense is structurally broken, and the fix has to come from how defenses are packaged, not from better RepE losses.*

**Comment by kabachuha** [comment]
Can we do a full rank fine-tune (no LoRA) with these techniques on a smaller model (e.g. 2B LLM), trainable at home? I did full tunes of 2B models at home on a 5090, with Unsloth. (With a second GPU we can perfectly go to 4B) Or are the 2B models too stupid to learn meaningful refusal behaviors?

**Comment by kabachuha** [comment]
Are you always assuming W_base is available? Maybe we can just put the model as a virtual black-box, to prevent the cheating logic, and try some actual methods?

**Comment by wuwangzhang1216** [comment]
Let me be precise about what our result does and doesn't claim, because I think we're talking past each other.

**What we claim.** Three recent "anti-abliteration" releases — DeepRefusal, Mistral-7B-Instruct-RR, Llama-3-8B-Instruct-RR — ship their safety circuit as a rank-16 LoRA-shaped delta on top of a **publicly available base model**. Under that specific packaging, the defense is structurally removable by arithmetic on published artifacts: `W_defended − W_base`, SVD, lerp to `λ=0.0`, standard abliteration on the recovered base. That's the result. It's scoped, it's reproducible, and it doesn't require any assumption that isn't already public.

**What we do not claim.** We are not claiming to have defeated the RR / DeepRefusal *training objective*. Those objectives are about making the refusal direction robust to pure activation-level attacks, and the DeepRefusal paper's own Table 1 is consistent with the view that, against that threat model, they work: Refusal Ablation, Refusal-Transfer, GCG, and Prefilling all report ≤2% ASR on their checkpoint. I have no reason to dispute those numbers and our work does not. Our finding is one level up the stack: **the training objective can be sound and the released artifact can still be trivially broken, because the packaging leaks the safety delta in extractable form.** Those are two separate claims and conflating them is what makes "cheating" sound plausible when it isn't.

**On the `virtual black-box` framing.** It's a legitimate research setting, but it's a *different* one from the one the published defenses actually ship under. Asking us to pretend `W_base` is unavailable when the defender has literally published it on Hugging Face and named it in their README reframes the problem into one the defender didn't commit to. If the defenders want the black-box threat model to be the operative one, the fix is on their side: don't release the hardened weights, or release a distilled-from-scratch student with no extractable delta. Until then, `W_base` isn't a cheat code — it's public infrastructure, and any realistic adversary has it.

**The sharp version of our claim, stated once:** for any defense that (a) is trained as a fine-tune from a public base and (b) releases the post-fine-tune weights, the defender has published the safety circuit in a form that is recoverable by SVD regardless of how clever the training loss was. This is a statement about release policy, not about RepE. It is falsified the moment someone shows a released checkpoint in this category where `W_defended − W_base` lacks exploitable low-rank structure. We have not yet seen one; if you know of one, that's the interesting pointer.
